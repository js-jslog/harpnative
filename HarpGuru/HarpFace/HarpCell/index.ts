export { HarpCell } from './components'
export type { HarpCellProps, YXCoord } from './types'
