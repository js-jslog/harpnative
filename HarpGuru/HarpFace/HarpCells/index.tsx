import React from 'react'

import type { HarpRowProps } from '../HarpRow'
import { HarpCell } from '../HarpCell'
import type { YXCoord } from '../HarpCell'

export const getHarpCells = (props: HarpRowProps): React.ReactElement[] => {
  const { yCoord, ...harpCellPropsPart } = props
  const { xRange } = harpCellPropsPart

  const harpCells = xRange.map((xCoord) => {
    const yxCoord: YXCoord = [yCoord, xCoord]
    return <HarpCell key={xCoord} {...harpCellPropsPart} yxCoord={yxCoord} />
  })

  return harpCells
}
