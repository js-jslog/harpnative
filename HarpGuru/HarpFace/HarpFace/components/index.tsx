import {StyleSheet, View} from 'react-native'
import React from 'react'

import type {HarpFaceProps} from '../types'
import { HoleNumberRow } from '../../HoleNumberRow'
import { getHarpRows } from '../../HarpRows'

const styles = StyleSheet.create({
  face: {
    flex: 5,
    flexDirection: 'column',
    justifyContent: 'space-between',
    alignItems: 'stretch',
  },
})

export const HarpFace = (props: HarpFaceProps): React.ReactElement => {
  const harpRows = getHarpRows(props)
  return (
    <View style={styles.face}>
      { harpRows.top }
      <HoleNumberRow {...props} />
      { harpRows.bottom }
    </View>
  )
}
