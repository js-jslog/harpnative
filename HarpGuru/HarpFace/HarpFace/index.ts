export { HarpFace } from './components'
export { getHarpFaceFacts } from './getHarpFaceFacts'
export { columnWidth, rowHeight } from './styles'
