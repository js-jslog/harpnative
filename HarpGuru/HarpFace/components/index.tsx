import React from 'react';
import {StyleSheet, Text, View} from 'react-native';

import {HarpRow} from '../../HarpRow';

const styles = StyleSheet.create({
  body: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
});

export const HarpFace = () => {
  return (
    <>
      <HarpRow />
      <View style={styles.body}>
        <Text>Try editing me! 🎉</Text>
      </View>
    </>
  );
};
