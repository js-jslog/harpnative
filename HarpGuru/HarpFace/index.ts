export { HarpFace } from './HarpFace'
export { DisplayModes } from './types'
export type { HarpFaceProps } from './types'
