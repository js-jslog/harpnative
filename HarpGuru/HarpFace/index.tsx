import React from 'react';
import {StyleSheet, Text, View} from 'react-native';

const styles = StyleSheet.create({
  body: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
});

export const HarpFace = () => {
  return (
    <View style={styles.body}>
      <Text>Try editing me! 🎉</Text>
    </View>
  );
};
