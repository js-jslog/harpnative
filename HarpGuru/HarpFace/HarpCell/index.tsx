import {View, StyleSheet, Button} from 'react-native'
import React from 'react'
import { getHarpStrata } from 'harpstrata'
import type { HarpStrataProps } from 'harpstrata'

import { DisplayModes } from '../HarpFace'

import { getToggledActiveDegreeIds } from './getToggledActiveDegreeIds'
import { analysePosition } from './analysePosition'
import type { PositionFacts } from './analysePosition'

import {HarpCellProps} from './types'

const styles = StyleSheet.create({
  cell: {
    backgroundColor: 'skyblue',
    justifyContent: 'center',
    alignItems: 'center',
    width: 30,
    height: 30,
  },
})

export const HarpCell = (props: HarpCellProps): React.ReactElement => {
  const positionFacts: PositionFacts = analysePosition(props)
  const { setActiveHarpStrata, harpStrata, displayMode } = props
  const { thisDegree, thisPitch } = positionFacts
  const { id: degreeId } = thisDegree || { id: undefined }
  const { id: pitchId } = thisPitch || { id: undefined }

  const displayValue = (displayMode === DisplayModes.Degree ? degreeId : pitchId)

  const { apparatus: { id: apparatusId }, pozitionId, harpKeyId, isActiveComplex: { activeDegreeIds }} = harpStrata

  const toggleActiveIdsIfHoleExists = (): void => {
    if (degreeId === undefined) return

    const newActiveDegreeIds = getToggledActiveDegreeIds(degreeId, activeDegreeIds)
    const activeIds = newActiveDegreeIds

    const newHarpStrataProps: HarpStrataProps = {
      apparatusId, pozitionId, harpKeyId, activeIds
    }
    const newHarpStrata = getHarpStrata(newHarpStrataProps)
    setActiveHarpStrata(newHarpStrata)
  }

  const accessibleContent = 
    <View accessible={true} accessibilityRole='button' style={styles.cell}>
      <Button onPress={(): void => toggleActiveIdsIfHoleExists()} title={displayValue || ''} />
    </View>

  const inAccessibleContent = 
    <View accessible={false} style={styles.cell}>
    </View>

  const content = (thisDegree === undefined ? inAccessibleContent : accessibleContent)

  return content
}

export type { YXCoord } from './types'
