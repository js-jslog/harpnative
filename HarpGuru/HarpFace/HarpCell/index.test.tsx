import React from 'react'
import { PitchIds, DegreeIds } from 'harpstrata'
import {render, fireEvent} from '@testing-library/react-native'

import { DisplayModes } from '../HarpFace'
import { exampleHarpFaceProps } from '../HarpFace'

import {HarpCell} from './index'

jest.mock('react-native/Libraries/Components/Touchable/TouchableOpacity', () => 'TouchableOpacity')

test('A component is rendered with the Degree or Pitch value in its text view depending on the DisplayMode selected', () => {
  const harpFaceProps = { ...exampleHarpFaceProps, activeDisplayMode: DisplayModes.Degree }
  const {getByText, rerender} = render(<HarpCell {...harpFaceProps} yxCoord={[3,0]} />)

  expect(getByText(DegreeIds.Second)).toBeTruthy()

  rerender(<HarpCell {...harpFaceProps} activeDisplayMode={DisplayModes.Pitch} yxCoord={[3,0]} />)

  expect(getByText(PitchIds.D)).toBeTruthy()
})

test('A component is rendered with an a11y role of button', () => {
  const harpFaceProps = { ...exampleHarpFaceProps, activeDisplayMode: DisplayModes.Degree }
  const { getByRole } = render(<HarpCell {...harpFaceProps} yxCoord={[3,0]} />)

  expect(getByRole('button')).toBeTruthy()
})

test('A component is rendered without an a11y role of button if it has no content', () => {
  const harpFaceProps = { ...exampleHarpFaceProps, activeDisplayMode: DisplayModes.Degree }
  const { queryByRole  } = render(<HarpCell {...harpFaceProps} yxCoord={[0,0]} />)

  expect(queryByRole('button')).toBeNull()
})

test('A press of the componenet results in toggled active ids in the harpstrata passed to the paramaterised setter', () => {
  const setActiveHarpStrata = jest.fn()
  const harpFaceProps = { ...exampleHarpFaceProps, activeDisplayMode: DisplayModes.Degree, setActiveHarpStrata }

  const { getByText } = render(<HarpCell {...harpFaceProps} yxCoord={[3,0]} />)

  fireEvent.press(getByText(DegreeIds.Second))

  const { mock: { calls: [[ newHarpStrata ]]}} = setActiveHarpStrata
  const { isActiveComplex: { activeDegreeIds: newDegreeIds, activePitchIds: newPitchIds }} = newHarpStrata

  const expectedDegreeIds = [DegreeIds.Second]
  const expectedPitchIds = [PitchIds.D]

  expect(setActiveHarpStrata.mock.calls.length).toBe(1)  
  expect(newDegreeIds).toStrictEqual(expectedDegreeIds)
  expect(newPitchIds).toStrictEqual(expectedPitchIds)
})

test('A snapshot of a populated cell', () => {
  const setActiveHarpStrata = jest.fn()
  const harpFaceProps = { ...exampleHarpFaceProps, activeDisplayMode: DisplayModes.Degree, setActiveHarpStrata }

  const { container } = render(<HarpCell {...harpFaceProps} yxCoord={[3,0]} />)
  expect(container).toMatchSnapshot()
})

test('A snapshot of an empty cell', () => {
  const setActiveHarpStrata = jest.fn()
  const harpFaceProps = { ...exampleHarpFaceProps, activeDisplayMode: DisplayModes.Degree, setActiveHarpStrata }

  const { container } = render(<HarpCell {...harpFaceProps} yxCoord={[0,0]} />)
  expect(container).toMatchSnapshot()
})
