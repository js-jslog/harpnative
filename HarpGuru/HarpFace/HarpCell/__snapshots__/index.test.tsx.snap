// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A snapshot of a populated cell 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <TouchableOpacity
    accessibilityRole="button"
    accessible={true}
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#fefefe",
          "borderColor": "#ced4da",
          "borderRadius": 7.5394000000000005,
          "borderWidth": 1.1,
          "elevation": 0,
          "height": 31.9374,
          "justifyContent": "center",
          "width": 31.9374,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#ced4da",
            "fontSize": 12.199000000000002,
          }
        }
      >
        2
      </Text>
    </View>
  </TouchableOpacity>
</View>
`;

exports[`A snapshot of an empty cell 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <TouchableOpacity
    accessible={false}
    disabled={true}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#fefefe",
        "borderColor": "#ced4da",
        "borderRadius": 7.5394000000000005,
        "borderWidth": 0,
        "elevation": 0,
        "height": 31.9374,
        "justifyContent": "center",
        "width": 31.9374,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#fefefe",
          "borderColor": "#ced4da",
          "borderRadius": 7.5394000000000005,
          "borderWidth": 0,
          "elevation": 0,
          "height": 31.9374,
          "justifyContent": "center",
          "width": 31.9374,
        }
      }
    />
  </TouchableOpacity>
</View>
`;
